lib/tumblr.js

```diff
@@ -1,262 +1,452 @@
+    },
+    POST: {
+        // Post authoring
+        createPost: '/blog/:blogIdentifier/post',
+        editPost: '/blog/:blogIdentifier/post/edit',
+        reblogPost: '/blog/:blogIdentifier/post/reblog',
+        deletePost: ['/blog/:blogIdentifier/post/delete', ['id']],
+        // Blog interaction
+        followBlog: ['/user/follow', ['url']],
+        unfollowBlog: ['/user/unfollow', ['url']],
+        // Post interaction
+        likePost: ['/user/like', ['id', 'reblog_key']],
+        unlikePost: ['/user/unlike', ['id', 'reblog_key']],
+    },
 };
 
-var baseURL = 'http://api.tumblr.com/v2';
-var version = require('../package.json').version;
+/**
+ * ## forceFullBlogUrl
+ *
+ * Turns a blog name to a full blog URL
+ *
+ * @param  {String} blogUrl: blog name or URL
+ *
+ * @return {String} full blog URL
+ */
+function forceFullBlogUrl(blogUrl) {
+    if (blogUrl.indexOf('.') < 0) {
+        blogUrl += '.tumblr.com';
+    }
+    return blogUrl;
+}
 
-var calls = {
+/**
+ * ## createFunction
+ *
+ * Creates a named function with the desired signature
+ *
+ * @param  {String} name: function name
+ * @param  {Array} [args]: array of argument names
+ * @param  {Function} fn: function that contains the logic that should run
+ *
+ * @return {Function} a named function that takes the desired arguments
+ */
+function createFunction(name, args, fn) {
+    if (typeof args === 'function') {
+        fn = args;
+        args = [];
+    }
 
-  postCreation: function (type, requireOptions, acceptsData) {
-    return function (blogName, options, callback) {
-      requireValidation(options, requireOptions);
-      options.type = type;
-      if (!acceptsData) {
-        delete options.data;
-      }
-      this._post(blogPath(blogName, '/post'), options, callback);
-    };
-  },
+    return new Function('body',
+        'return function ' + name + '(' + args.join(', ') + ') { return body.apply(this, arguments); };'
+    )(fn);
+}
 
-  getWithOptions: function (path) {
-    return function (options, callback) {
-      if (isFunction(options)) { callback = options; options = {}; }
-      this._get(path, options, callback);
-    };
-  },
+/**
+ * ## requestCallback
+ *
+ * Wraps a function for use as a request callback
+ *
+ * @param  {Function} callback: function to wrap
+ *
+ * @return {Function} request callback
+ */
+function requestCallback(callback) {
+    if (!callback) {
+        return undefined;
+    }
 
-  blogList: function (path) {
-    return function (blogName, options, callback) {
-      if (isFunction(options)) { callback = options; options = {}; }
-      this._get(blogPath(blogName, path), options, callback);
+    return function(err, response, body) {
+        if (err) {
+            return callback(err);
+        }
+
+        if (response.statusCode >= 400) {
+            var errString = body.meta ? body.meta.msg : body.error;
+            return callback(new Error('API error: ' + response.statusCode + ' ' + errString));
```

Comment:
Can the response object be provided to the callback as a third argument? Having the ability to retrieve information from it could be valuable for handling errors:

```js
var blogName = 'my-awesome-blog.tumblr.com';
var params = {
    title: "my 8 billionth post today!!",
    body: "I'm going for a record"
};

client.createTextPost(blogName, params, function(err, data, response) {
    if (err) {
        if (response.statusCode == 429) {
            var limitResetSeconds = response.headers['X-RateLimit-Reset'];
            enqueueRequest(response.request, limitResetSeconds);
        }
    }
});
```